File: 733-flood-fill/733-flood-fill.cpp
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int newCol) {
        int col = image[sr][sc];
        if (col != newCol)
            dfs(image, sr, sc, col, newCol);
        return image;
    }
    
    void dfs(vector<vector<int>>& image, int r, int c, int col, int newCol) {
        if (image[r][c] == col) {
            image[r][c] = newCol;
            
            if (r >= 1)
                dfs (image, r - 1, c, col, newCol);
            if (c >= 1)
                dfs (image, r, c - 1, col, newCol);
            if (r + 1 < image.size())
                dfs (image, r + 1, c, col, newCol);
            if (c + 1 < image[0].size())
                dfs (image, r, c + 1, col, newCol);
        }
    }
};

File: 733-flood-fill/733-flood-fill.java
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        int org = image[sr][sc];
        if (org == color)
            return image;
        helper(image, sr, sc, org, color);
        return image;
    }
    
    private void helper(int[][] image, int r, int c, int org, int color) {
        if (r < 0 || r >= image.length || c < 0 || c >= image[0].length || image[r][c] != org)
            return;
        image[r][c] = color;
        helper(image, r + 1, c, org, color);
        helper(image, r - 1, c, org, color);
        helper(image, r, c + 1, org, color);
        helper(image, r, c - 1, org, color);
    }
}
